### PR TITLE
[Kernel] support dynamic_validate_inputs for kv_cache_update

### PR DESCRIPTION
**Pull request by yaochengji**
# Description

Added the `dynamic_validate_inputs` argument in the kv_cache_update kernel.

# Tests
```
pytest -s -v tests/ragged_kv_cache_update_test.py -k test_basic
```

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

**Comment by yaochengji** [comment]
@bythew3i could you take a look?

@sixiang-google , you might enable this check when debugging.

File: tpu_commons/kernels/ragged_kv_cache_update.py [hunk 1]
```diff
@@ -67,6 +67,23 @@ def _kv_cache_update_kernel(
...
+        kv_cache_start = slices[0, i]
+        new_kv_start = slices[1, i]
+        slice_len = slices[2, i]
+        if kv_cache_start >= kv_cache_token_num:
+            raise ValueError(
+                f"{kv_cache_start=} must be less than {kv_cache_token_num=}")
+        if new_kv_start >= new_token_num:
+            raise ValueError(
+                f"{new_kv_start=} must be less than {new_token_num=}")
+        if slice_len > page_size:
```
Comment (bythew3i) [inline tpu_commons/kernels/ragged_kv_cache_update.py]: I would suggest using this pattern:

if not (0 < slice_len <= page_size)... so you can cover all the OOB check.

File: tpu_commons/kernels/ragged_kv_cache_update.py [hunk 2]
```diff
@@ -67,6 +67,23 @@ def _kv_cache_update_kernel(
         async_copy.wait()
 
 
+def _dynamic_validate_inputs(slices, new_token_num, kv_cache_token_num,
+                             page_size):
+    for i in range(slices.shape[1]):
+        kv_cache_start = slices[0, i]
+        new_kv_start = slices[1, i]
+        slice_len = slices[2, i]
+        if kv_cache_start >= kv_cache_token_num:
```
Comment (bythew3i) [inline tpu_commons/kernels/ragged_kv_cache_update.py]: You may also want to check these **invalid** cases:

- `kv_cache_start < 0`
- `kv_cache_start  + slice_len >= kv_cache_token_num`
- `new_kv_start < 0`
- `new_kv_start + slice_len >= new_token_num`

Also I think you may need check this **invalid** case (to make sure the slice update is within same page):
- `new_kv_start // page_size != (new_kv_start + slice_len) // page_size`
- `kv_cache_start // page_size != (kv_cache_start + slice_len) // page_size`

Comment (yaochengji) [reply tpu_commons/kernels/ragged_kv_cache_update.py]: Great suggestions, Jevin! Do you think we need to make sure the token indices to be updated needs to cover the front part of the new_kv_cache tensor with no bubbles and no overlap? Although I don't think it will trigger crash in the kernel.

File: tpu_commons/kernels/ragged_kv_cache_update.py (hunk 1 above)
Comment (yaochengji) [reply tpu_commons/kernels/ragged_kv_cache_update.py]: Thanks Jevin! I think it should be `0 <= slice_len <= page_size` because the padding's slice_len is 0.

File: tpu_commons/kernels/ragged_kv_cache_update.py [hunk 3]
```diff
@@ -67,6 +67,23 @@ def _kv_cache_update_kernel(
         async_copy.wait()
 
 
+def _dynamic_validate_inputs(slices, new_token_num, kv_cache_token_num,
+                             page_size):
+    for i in range(slices.shape[1]):
```
Comment (bythew3i) [inline tpu_commons/kernels/ragged_kv_cache_update.py]: If it is dynamic check we can just use `num_slices`?

File: tpu_commons/kernels/ragged_kv_cache_update.py (hunk 1 above)
Comment (bythew3i) [reply tpu_commons/kernels/ragged_kv_cache_update.py]: IIUC, if we iterate through `num_slices` instead of `slices.shape[1]` we won't check padding.

File: tpu_commons/kernels/ragged_kv_cache_update.py (hunk 2 above)
Comment (bythew3i) [reply tpu_commons/kernels/ragged_kv_cache_update.py]: Did you mean checking if there is any overlap between slices?

**Comment by yaochengji** [comment]
> Did you mean checking if there is any overlap between slices?

Yeah, no overlap and only one bubble in the end.

File: tpu_commons/kernels/ragged_kv_cache_update.py (hunk 2 above)
Comment (yaochengji) [reply tpu_commons/kernels/ragged_kv_cache_update.py]: > `new_kv_start // page_size != (new_kv_start + slice_len) // page_size`

I'm afraid this condition might not be satisfied even in a valid case. E.g. page_size = 16, 
slice0: len = 3
slice1: len = 16

Comment (yaochengji) [reply tpu_commons/kernels/ragged_kv_cache_update.py]: > `kv_cache_start // page_size != (kv_cache_start + slice_len) // page_size`

should be `kv_cache_start // page_size != (kv_cache_start + slice_len - 1) // page_size`?

File: tpu_commons/kernels/ragged_kv_cache_update.py (hunk 3 above)
Comment (yaochengji) [reply tpu_commons/kernels/ragged_kv_cache_update.py]: I think we'd better also check the padding part because it will also go into the kernel, WDYT?

Comment (bythew3i) [reply tpu_commons/kernels/ragged_kv_cache_update.py]: IMO, we should not care about the padding. Because  I don't think we use the padding in the kernel.

File: tpu_commons/kernels/ragged_kv_cache_update.py (hunk 2 above)
Comment (bythew3i) [reply tpu_commons/kernels/ragged_kv_cache_update.py]: > should be kv_cache_start // page_size != (kv_cache_start + slice_len - 1) // page_size?

Yes!

File: tpu_commons/kernels/ragged_kv_cache_update.py (hunk 3 above)
Comment (yaochengji) [reply tpu_commons/kernels/ragged_kv_cache_update.py]: We will, even with dynamic grid.
E.g. num_slice_per_block=8, actual_slices_num=5, and we padded the slices_num to 16. There're 3 padding slices used in the kernel.

File: tests/ragged_kv_cache_update_test.py [hunk 4]
```diff
@@ -60,20 +60,24 @@ def _generate_data(self, page_size, combined_kv_head_num, head_dim,
...
         old_kv_cache_copy = kv_cache.copy()
 
-        updated_kv_cache = kv_cache_update(
-            new_kv,
-            slot_mapping,
-            kv_cache,
-            num_slices,
-            page_size=page_size,
-            num_slices_per_block=num_slices_per_block)
+        with jax.disable_jit(disable=dynamic_validate_inputs):
```
Comment (vanbasten23) [inline tests/ragged_kv_cache_update_test.py:72]: I wonder why in the service code we don't disable the JIT.

Comment (vanbasten23) [inline tests/ragged_kv_cache_update_test.py:72]: I assume the use case is that by default we set dynamic_validate_inputs to False. When there is a bug, then we set it to True, right?

File: tpu_commons/kernels/ragged_kv_cache_update.py (hunk 3 above)
Comment (bythew3i) [reply tpu_commons/kernels/ragged_kv_cache_update.py]: Ah I see, you did not guard that using either dynamic for loop or if-else branch. Then I think it is really dangerous, the kernel assumes padding are zero. If non-zero, it is actually going to emit DMA or some OOB access (eg, stupid large number in padding). I would suggest changing the python for to lax.for_i to make sure only effective DMA is issued.

Comment (yaochengji) [reply tpu_commons/kernels/ragged_kv_cache_update.py]: I tried to use `lax.for_i ` before, but it will trigger a jax trace error.

Currently, all the padding slice has slice_len=0, do you still think it's dangerous? Or we can add the check that when `i >= slice_num`, slice_len[i] = 0.

File: tests/ragged_kv_cache_update_test.py (hunk 4 above)
Comment (yaochengji) [reply tests/ragged_kv_cache_update_test.py:72]: Yes

Comment (yaochengji) [reply tests/ragged_kv_cache_update_test.py:72]: Because jax.jit is already annotated in the `kv_cache_update` method.

File: tpu_commons/kernels/ragged_kv_cache_update.py (hunk 3 above)
Comment (bythew3i) [reply tpu_commons/kernels/ragged_kv_cache_update.py]: > Currently, all the padding slice has slice_len=0

You can only guarantee that on call site. But kernel just assume 0. You can do sth similar to https://github.com/vllm-project/tpu_commons/blob/main/tpu_commons/kernels/ragged_paged_attention/kernel.py#L53 while also make the size to 0. In this way the padded start index is guranteed to be in valid range instead of assuming user has padded to 0 already.

File: tpu_commons/kernels/ragged_kv_cache_update.py (hunk 2 above)
Comment (yaochengji) [reply tpu_commons/kernels/ragged_kv_cache_update.py]: > kv_cache_start  + slice_len >= kv_cache_token_num
> new_kv_start + slice_len >= new_token_num

Should be `kv_cache_start  + slice_len > kv_cache_token_num` and `new_kv_start + slice_len > new_token_num`?

File: tpu_commons/kernels/ragged_kv_cache_update.py (hunk 3 above)
Comment (yaochengji) [reply tpu_commons/kernels/ragged_kv_cache_update.py]: Thanks for the suggestion! I remembered during the implementation of kernel, I was afraid it will trigger some overhead and that's why I left the input validation to the kernel users.

File: tpu_commons/kernels/ragged_kv_cache_update.py (hunk 2 above)
Comment (bythew3i) [reply tpu_commons/kernels/ragged_kv_cache_update.py]: yes, right!

**Comment by yaochengji** [comment]
@bythew3i @vanbasten23 

Thanks for your review. I think I already added all the checks @bythew3i you mentioned and additional checks to make sure the slices in new_kv is continuous, no overlap in kv cache slices.

A new test is also added to make sure the appropriate exceptions are raised.

Could you take another look?
